From 7f02b11938ee208d85223dc2cfbcd4e2543747b3 Mon Sep 17 00:00:00 2001
From: Jianxu Zheng <jian.xu.zheng@intel.com>
Date: Thu, 14 Jul 2016 22:33:33 +0800
Subject: [PATCH 10/11] media:entity: graph walk starting from pad

Add new parameter in media_entity_graph_walk_start()
Start walking the media graph at a given entity and pad

This function initializes the graph traversal structure to walk
the entities graph starting at the given entity and pad.
As some entity may be on the cross node on pipeline, so add one
parameter to start the graph traverse.

Change-Id: I546d259992eadbaf9d79aae42fbe5e6a8f12022a
Signed-off-by: Jianxu Zheng <jian.xu.zheng@intel.com>
Reviewed-on: https://android.intel.com/532952
Reviewed-by: Qiu, Tian Shu <tian.shu.qiu@intel.com>
Reviewed-by: Cao, Bingbu <bingbu.cao@intel.com>
Reviewed-by: Meng, WeiX <weix.meng@intel.com>
Tested-by: Meng, WeiX <weix.meng@intel.com>
Reviewed-by: Yang, Fan F <fan.f.yang@intel.com>
---
 drivers/media/media-entity.c | 19 +++++++++----------
 include/media/media-entity.h |  2 +-
 2 files changed, 10 insertions(+), 11 deletions(-)

diff --git a/drivers/media/media-entity.c b/drivers/media/media-entity.c
index d588d58..56bd6aa 100644
--- a/drivers/media/media-entity.c
+++ b/drivers/media/media-entity.c
@@ -150,7 +150,7 @@ static struct media_entity *stack_pop(struct media_entity_graph *graph)
 /**
  * media_entity_graph_walk_start - Start walking the media graph at a given entity
  * @graph: Media graph structure that will be used to walk the graph
- * @entity: Starting entity
+ * @pad: starting pad
  *
  * This function initializes the graph traversal structure to walk the entities
  * graph starting at the given entity. The traversal structure must not be
@@ -158,17 +158,17 @@ static struct media_entity *stack_pop(struct media_entity_graph *graph)
  * safely be freed.
  */
 void media_entity_graph_walk_start(struct media_entity_graph *graph,
-				   struct media_entity *entity)
+				   struct media_pad *pad)
 {
 	graph->top = 0;
 	graph->stack[graph->top].entity = NULL;
 	bitmap_zero(graph->entities, MEDIA_ENTITY_ENUM_MAX_ID);
 
-	if (WARN_ON(entity->id >= MEDIA_ENTITY_ENUM_MAX_ID))
+	if (WARN_ON(pad->entity->id >= MEDIA_ENTITY_ENUM_MAX_ID))
 		return;
 
-	__set_bit(entity->id, graph->entities);
-	stack_push(graph, entity, -1);
+	__set_bit(pad->entity->id, graph->entities);
+	stack_push(graph, pad->entity, pad->index);
 }
 EXPORT_SYMBOL_GPL(media_entity_graph_walk_start);
 
@@ -230,8 +230,7 @@ media_entity_graph_walk_next(struct media_entity_graph *graph)
 		 * Are the local pad and the pad we came from connected
 		 * internally in the entity ?
 		 */
-		if (from_pad != -1 &&
-		    !media_entity_has_route(entity, from_pad, local->index)) {
+		if (!media_entity_has_route(entity, from_pad, local->index)) {
 			link_top(graph)++;
 			continue;
 		}
@@ -279,7 +278,7 @@ __must_check int media_entity_pipeline_start(struct media_entity *entity,
 
 	mutex_lock(&mdev->graph_mutex);
 
-	media_entity_graph_walk_start(&graph, entity);
+	media_entity_graph_walk_start(&graph, &entity->pads[0]);
 
 	while ((entity = media_entity_graph_walk_next(&graph))) {
 		DECLARE_BITMAP(active, MEDIA_ENTITY_MAX_PADS);
@@ -354,7 +353,7 @@ error:
 	 * Link validation on graph failed. We revert what we did and
 	 * return the error.
 	 */
-	media_entity_graph_walk_start(&graph, entity_err);
+	media_entity_graph_walk_start(&graph, &entity_err->pads[0]);
 
 	while ((entity_err = media_entity_graph_walk_next(&graph))) {
 		entity_err->stream_count--;
@@ -394,7 +393,7 @@ void media_entity_pipeline_stop(struct media_entity *entity)
 
 	mutex_lock(&mdev->graph_mutex);
 
-	media_entity_graph_walk_start(&graph, entity);
+	media_entity_graph_walk_start(&graph, &entity->pads[0]);
 
 	while ((entity = media_entity_graph_walk_next(&graph))) {
 		entity->stream_count--;
diff --git a/include/media/media-entity.h b/include/media/media-entity.h
index cda9614..720a720 100644
--- a/include/media/media-entity.h
+++ b/include/media/media-entity.h
@@ -161,7 +161,7 @@ bool media_entity_has_route(struct media_entity *entity, unsigned int sink,
 			    unsigned int source);
 
 void media_entity_graph_walk_start(struct media_entity_graph *graph,
-		struct media_entity *entity);
+				   struct media_pad *pad);
 struct media_entity *
 media_entity_graph_walk_next(struct media_entity_graph *graph);
 __must_check int media_entity_pipeline_start(struct media_entity *entity,
-- 
2.9.2

